### PR TITLE
Adding the arcconf plugin support to libstoragemgmt

### PR DESCRIPTION
**Pull request by brraghav**
Adding the support of arcconf plugin into the libtoragemgmt.

### CODE REVIEW

**Comment by cathay4t** [comment]
@joehandzik I noticed the license line is HPE, could you review this PR also?

@brraghav Could you provide a brief information about supported hardware and steps of setup? Normally those information should be found in mange page `arcconf_lsmplugin (1)`.

**Comment by cathay4t** [comment]
@brraghav `arcconf` seems not fit for a plugin name, how about `aacraid`?

**Comment by brraghav** [comment]
@cathay4t  It's a good suggestion, but, there is an "aacraid" inbox driver for Adaptec controllers(Series 6/7/8). I just thought that having the same name for driver and plugin might be confusing. Well, will discuss with few here and come back quickly on this.

File: plugin/arcconf/arcconf.py [hunk 1]
```diff
@@ -1,5 +1,5 @@
 # Copyright (C) 2015-2016 Red Hat, Inc.
-# (C) Copyright 2015-2016 Hewlett Packard Enterprise Development LP
```
Comment (cathay4t) [inline plugin/arcconf/arcconf.py]: @brraghav You should not do that. If you copied code from hpsa plugin, you should keep the original license line and only add your new one.

File: plugin/arcconf/arcconf.py [hunk 2]
```diff
@@ -14,7 +14,6 @@
 # License along with this library; If not, see <http://www.gnu.org/licenses/>.
 #
 # Author: Gris Ge <fge@redhat.com>
-#         Joe Handzik <joseph.t.handzik@hpe.com>
```
Comment (cathay4t) [inline plugin/arcconf/arcconf.py]: @brraghav Why remove author? your code is based on hpsa plugin codes.

**Comment by cathay4t** [comment]
@brraghav This plugin is almost identical to hpsa plugin, except `_DEFAULT_BIN_PATHS`. Why are we doing this? Have you talk with @joehandzik about unifing the work?

File: plugin/arcconf/arcconf.py (hunk 2 above)
Comment (joehandzik) [inline plugin/arcconf/arcconf.py]: Agreed, just leave the author list the way it was, and append your name and contact information to it, @brraghav

**Comment by joehandzik** [comment]
@cathay4t Yes, I've emailed with @brraghav about what he's doing. After looking at the code, I don't know that what he's doing is 'almost identical' to the hpsa plugin, the aacraid configuration utility output is formatted differently (we'd need to put an extra abstraction layer in, which seems overkill, and then any changes to the plugin would need to be tested with two different tools).

File: plugin/arcconf/arcconf.py (hunk 2 above)
Comment (brraghav) [inline plugin/arcconf/arcconf.py]: Ok. Got it. Initially, I had done the same. 
I had appended my name under the author and had HPE in the copyright as is. I mistook the first comment about the HPE copyright. Sorry for that.

**Comment by brraghav** [comment]
@cathay4t I have the .travis.yml file in my repository and the repository is also activated in the Travis CI website. Is there anything else I need to do?

File: plugin/arcconf/arcconf.py [hunk 3]
```diff
@@ -0,0 +1,963 @@
...
+                if 'HardDrive' in decoded_json['Controller']['Channel'][channel]:
+                    hd_disks_num = len(decoded_json['Controller']['Channel'][channel]['HardDrive'])
+                    if hd_disks_num > 0:
+                        for hd_disk in range(hd_disks_num):
+                            rc_lsm_disks.append(Arcconf._arcconf_disk_to_lsm_disk(decoded_json['Controller']['Channel'][channel]['HardDrive'][hd_disk], sys_id, cntrl_num, flag_free=False))            
+
+
+        return search_property(rc_lsm_disks, search_key, search_value)
+
+    @_handle_errors
```
Comment (joehandzik) [inline plugin/arcconf/arcconf.py]: I would remove anything that returns "None" like this and remove support in the 'Capabilities' method for them. It's incorrect to label them as supported if they aren't yet (support can always be added later).

File: plugin/arcconf/arcconf.py [hunk 4]
```diff
@@ -0,0 +1,963 @@
...
+        try:
+            self._arcconf_exec(['delete', ctrl_num, 'logicaldrive', array_num], COMPONENT[5], flag_convert=False, flag_force=True)
+        except ExecError:
+            raise LsmError(
+                ErrorNumber.NOT_FOUND_VOLUME,
+                "Volume not found")
+
+        return None
+
+    def volume_enable(self, volume, flags=Client.FLAG_RSVD):
```
Comment (joehandzik) [inline plugin/arcconf/arcconf.py]: Again for this chunk, remove if not supported and update Capabilities accordingly.

File: plugin/arcconf/arcconf.py [hunk 5]
```diff
@@ -0,0 +1,963 @@
...
+                        check_line = output_lines[line_num - 2]
+                        if check_line.find("Type") > 0:
+                            continue
+                        else:
+                            physical_data.update({temp_list[0] : temp_list[1]})
+                    else:
+                        physical_data.update({temp_list[0] : temp_list[1]})
+                    #if temp_list[0] == 'Drive Unique ID':
+                    if temp_list[0] == 'SSD':
+                        #list_data.append(list(physical_data), physical_data[0])
```
Comment (brraghav) [inline plugin/arcconf/arcconf.py]: Ok. Deleted the unused code in patch-3 commit.

File: plugin/arcconf/arcconf.py [hunk 6]
```diff
@@ -0,0 +1,963 @@
...
+
+    return _ARCCONF_RAID_LEVEL_CONV.get(arcconf_raid_level, Volume.RAID_TYPE_UNKNOWN)
+
+
+def _lsm_raid_type_to_arcconf(raid_type):
+    try:
+        if raid_type == '0':
+            return "simple_volume"
+        return "simple_volume"
+        #return _LSM_RAID_TYPE_CONV[raid_type]
```
Comment (brraghav) [inline plugin/arcconf/arcconf.py]: removed unused code in the patch-3 commit

File: plugin/arcconf/arcconf.py (hunk 3 above)
Comment (brraghav) [inline plugin/arcconf/arcconf.py]: Removed all the functions that were not yet implemented. Removed from the capability list as well.

File: plugin/arcconf/arcconf.py [hunk 7]
```diff
@@ -0,0 +1,963 @@
+# Copyright (C) 2015-2016 Red Hat, Inc.
+# (C) Copyright 2015-2016 Hewlett Packard Enterprise Development LP
+# (C) Copyright 2017-2018 Microsemi Corporation
```
Comment (brraghav) [inline plugin/arcconf/arcconf.py]: Changed this to 2016-2017.

File: plugin/arcconf/arcconf.py [hunk 8]
```diff
@@ -0,0 +1,963 @@
...
+        # TODO(Gris Ge): Try degrade a RAID or fail a RAID.
+        return Pool.STATUS_OTHER, arcconf_array['Status']
+
+
+def _pool_id_of(sys_id, array_name):
+    return "%s:%s" % (sys_id, array_name.replace(' ', ''))
+
+
+def _disk_type_of(arcconf_disk):
+    disk_interface = arcconf_disk['interfaceType']
```
Comment (brraghav) [inline plugin/arcconf/arcconf.py]: Yes, they are integers. I will try to define constants with meaningful names. The enum values for the disk interface given by arcconf is different from that of the lsmcli..

**Comment by brraghav** [comment]
Created a new pull request with all review comments incorporated.

**Pull request closed without merging**